src/v/storage/index_state.cc

```diff
@@ -45,22 +46,48 @@ bool index_state::maybe_index(
       *this);
 
     bool retval = false;
+
+    // The first non-config batch in the segment, use its timestamp
+    // to override the timestamps of any config batch that was indexed
+    // by virtue of being the first in the segment.
+    if (user_data && non_data_timestamps) {
+        vassert(relative_time_index.size() == 1, "");
```

Comment:
i'm having trouble seeing how this assertion will always be true.

the first time through this we hit the `empty() == true` case and `retval = true` will then cause `add_entry` to run so relative_time_index.size will be 1.

but then for the next batch, if user_data is false, doesn't something have to prevent add_entry from being executed again prior user_data being true for some batch? 

EDIT:

nevermind, I think i see now that for this conditional protecting add_entry, `if ((accumulator >= step && user_data) || retval) {` retval won't ever be true (since the index is now non-empty) and user_data being true would have implied that we'd nice this case with assert.

lgtm